library/src/main/java/org/apache/apex/malhar/lib/state/spillable/TimeBasedPriorityQueue.java

```diff
@@ -115,8 +115,17 @@ public int compareTo(TimeWrapper<T> timeWrapper)
       } else if (this.timestamp > timeWrapper.getTimestamp()) {
         return 1;
       }
-
-      return 0;
+      
+      /**
+       * NOTE: the following use the equals() to implement the compareTo() for key.
+       * it should be OK as the compareTo() only used by TimeBasedPriorityQueue.sortedTimestamp, 
+       * which only care about the order of time ( the order for key doesn't matter ).
+       * But would cause problem if add other function which depended on the order of the key.
+       * 
+       * Add compare by hashCode when not equals in order to compatible with the interface for most cases.
+       * Anyway, the order of key is not guaranteed. And we should not return 0 if not equals
+       */
+      return key.equals(timeWrapper.key) ? 0 : (hashCode() - timeWrapper.hashCode() <= 0 ? -1 : 1);
```

Comment:
What's the implication of key order not being guaranteed? Managed state does guarantee key order.